### PR TITLE
Airlock data cleanup deletion when moving between stages

### LINKED ISSUE #1963: Airlock - data cleanup (deletion) when moving between stages
## Description

As a TRE Developer
I want data to be deleted from the source location after it has been copied to the destination location
So that data which is no longer needed will not accumulate over time (cost, cleanup)

## Acceptance criteria

- [ ] Deletion is triggered when the copy is done via a dedicated event
- [ ] Deletion does not block any other operation
- [ ] Airlock processor should handle the deletion

### PR DESCRIPTION
**Pull request by tanya-borisova**
# Resolves #1963 

## What is being addressed

Clean up data when moving between stages: after source blob has been copied into destination blob, source blob can be deleted. 

## How is this addressed

- Created a new Service Bus queue and EventGrid topic for blobs that need to be deleted
- Send an event from BlobCreatedTrigger function
- Added a new function ToDeleteTrigger that is triggered by messages in said queue
- Added "copied_from" field to blob metadata to trace where the blob has been copied from

In addition to this, some small refactorings (I left comments to the PR)

### CODE REVIEW

File: airlock_processor/shared_code/blob_operations.py [hunk 1]
```diff
@@ -29,25 +42,21 @@ def copy_data(source_account_name: str, destination_account_name: str, request_i
                                                    credential=credential)
     source_container_client = source_blob_service_client.get_container_client(container_name)
 
-    try:
```
Comment (tanya-borisova) [inline airlock_processor/shared_code/blob_operations.py:32]: I've removed this try-catch block as the request ID will still be logged on lines 51 and 57

File: airlock_processor/tests/shared_code/blob_operations.py [hunk 2]
```diff
@@ -1,31 +0,0 @@
...
-from azure.mgmt.storage import StorageManagementClient
-from azure.mgmt.storage.v2021_09_01.aio.operations import StorageAccountsOperations
-from azure.mgmt.storage.v2021_09_01.models import StorageAccountListKeysResult, StorageAccountKey
-from azure.storage.blob import BlobServiceClient
-
-from shared_code.blob_operations import create_container
-
-
-class TestBlobOperations(unittest.TestCase):
-    @patch('shared_code.blob_operations.get_blob_client_by_rg_and_account', return_value=Mock(BlobServiceClient))
```
Comment (tanya-borisova) [inline airlock_processor/tests/shared_code/blob_operations.py:13]: This test was never executed as the filename doesn't start with `test_`. It got stale and the function being imported no longer exists

**Review by tamirkamara — COMMENTED** [review]
LGTM.
Will let @anatbal do the final approval.

File: airlock_processor/ToDeleteTrigger/__init__.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Deleting blob
+    logging.info(f'Deleting blob {blob_name}...')
+    blob_client = container_client.get_blob_client(blob_name)
+    blob_client.delete_blob()
+
+    if blobs_num == 1:
+        # Need to delete the container too
+        logging.info(f'There was one blob in the container. Deleting container {container_name}...')
+        container_client.delete_container()
```
[the remark this answers is not in the document]
Comment (tanya-borisova) [reply airlock_processor/ToDeleteTrigger/__init__.py:31]: Good point, and will be a bit easier to read, too

File: airlock_processor/ToDeleteTrigger/__init__.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    storage_account_name, container_name, blob_name = blob_operations.get_blob_info_from_blob_url(blob_url=blob_url)
+    credential = blob_operations.get_credential()
+    blob_service_client = BlobServiceClient(
+        account_url=blob_operations.get_account_url(storage_account_name),
+        credential=credential)
+    container_client = blob_service_client.get_container_client(container_name)
+
+    # If it's the only blob in the container, we need to delete the container too
+    # Check how many blobs are in the container (note: this exausts the generator)
+    blobs_num = sum(1 for _ in container_client.list_blobs())
```
[the remark this answers is not in the document]
Comment (tanya-borisova) [reply airlock_processor/ToDeleteTrigger/__init__.py:20]: Good point! Then I don't need to worry about this potentially taking a long time if we ever have many blobs

File: airlock_processor/ToDeleteTrigger/__init__.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import logging
+import json
+
+import azure.functions as func
+from azure.storage.blob import BlobServiceClient
+
+from shared_code import blob_operations
+
+
+def delete_blob_and_container_if_last_blob(blob_url: str):
```
[the remark this answers is not in the document]
Comment (tanya-borisova) [reply airlock_processor/ToDeleteTrigger/__init__.py:10]: Sorry, which class do you mean? Do you mean blob_operations module?

[the remark this answers is not in the document]
Comment (tamirkamara) [reply airlock_processor/ToDeleteTrigger/__init__.py:10]: yes

File: airlock_processor/_version.py [hunk 6]
```diff
@@ -1 +1 @@
-__version__ = "0.4.2"
+__version__ = "0.4.294"
```
Comment (anatbal) [inline airlock_processor/_version.py:1]: I think this was left here by mistake?

**Review by anatbal — APPROVED** [review]
Looks good!

File: airlock_processor/ToDeleteTrigger/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tanya-borisova) [reply airlock_processor/ToDeleteTrigger/__init__.py:10]: Sorry for a delayed reply. I thought of that and I would say that maybe we should keep code in blob_operations that is reused in airlock_processor? But we don't do blob cleanup anywhere else currently so it's unlikely to be reused. What do you think?

[the remark this answers is not in the document]
Comment (tamirkamara) [reply airlock_processor/ToDeleteTrigger/__init__.py:10]: ok

**Comment by tanya-borisova** [comment]
/test-force-approve

The airlock tests pass, the only test that failed is test_create_guacamole_service_into_base_workspace but I think there are now intermittent issues with private endpoint which causes the test to fail often, so I won't attempt to get it to pass on this PR

**Pull request merged**